File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/CallResultConverter.java
package com.roxiemobile.networkingapi.network.rest;

import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity;
import com.roxiemobile.networkingapi.network.rest.response.error.nested.ConversionException;

import org.jetbrains.annotations.NotNull;

public interface CallResultConverter<Ti, To> {

    /**
     * Converts result from one format to another.
     */
    @NotNull CallResult<To> convert(@NotNull CallResult<Ti> result);

    /**
     * Converts response entity from one format to another.
     */
    @NotNull ResponseEntity<To> convert(@NotNull ResponseEntity<Ti> entity) throws ConversionException;
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/CallResultConverter.kt
package com.roxiemobile.networkingapi.network.rest

import com.roxiemobile.networkingapi.network.rest.response.ResponseEntity
import com.roxiemobile.networkingapi.network.rest.response.error.nested.ConversionException

interface CallResultConverter<Ti, To> {

// MARK: - Methods

    /**
     * Converts result from one format to another.
     */
    fun convert(callResult: CallResult<Ti>): CallResult<To>

    /**
     * Converts response entity from one format to another.
     */
    @Throws(ConversionException::class)
    fun convert(responseEntity: ResponseEntity<Ti>): ResponseEntity<To>
}
